GameRoom.ts

```diff
@@ -1,500 +1,637 @@
-                    message: `Player: ${this.state.playerList[player.loginName].displayName} left the game.`,
-                    origin: 'SERVER'
-                });
-                this.state.removePlayer(player.loginName);
-            }
-            player.gracePeriodTimeout = global.setTimeout(((p: Player) => {
-                if (!p.isConnected) {
-                    p.hasLeft = true;
-                    const pObj: PhysicsObjectState = this.state.physicsState.objects[p.figureId];
-                    pObj.setDisabled(true);
-                }
-                p.gracePeriodTimeout = undefined;
-            }).bind(this), 120000, player);
-        }
-        return undefined;
+    let joinedMsg = '';
+    if (isNewPlayer) {
+      joinedMsg = `${
+        this.state.playerList[options.login].displayName
+      } joined the game`;
+      player.figureId = this.state.physicsState.addPlayerFigure();
+    } else {
+      joinedMsg = `${
+        this.state.playerList[options.login].displayName
+      } reconnected to the game`;
+
+      // remove potential timeout
+      if (player.gracePeriodTimeout !== undefined) {
+        global.clearTimeout(player.gracePeriodTimeout);
+        player.gracePeriodTimeout = undefined;
+      }
+
+      // re-enable figure old game
+      const pObj: PhysicsObjectState = this.state.physicsState.objects[
+        player.figureId
+      ];
+      pObj.setDisabled(false);
+
+      // register new sessionId to old player object
+      player.clientId = client.id;
+      player.hasLeft = false;
     }
 
-    onChatMessage(client: Client, data: WsData) {
-        const player = this.state.getPlayerByClientId(client.id);
-        if (player !== undefined && data.type === MessageType.CHAT_MESSAGE) {
-            const msg: ChatMessage = {
-                type: MessageType.CHAT_MESSAGE,
-                message: data.message,
-                authorLoginName: player.loginName
-            };
-            this.broadcast(msg.type, msg);
+    player.isConnected = true;
+    this.broadcast(MessageType.JOIN_MESSAGE, {
+      type: MessageType.JOIN_MESSAGE,
+      message: joinedMsg,
+    });
+    return undefined;
+  }
+
+  onLeave(client: Client, consented?: boolean): void | Promise<any> {
+    const player = this.state.getPlayerByClientId(client.id);
+
+    const minutesToAdd =
+      (new Date().getTime() - player.joined.getTime()) / 60000;
+    UserController.addPlaytime(player.loginName, minutesToAdd);
+
+    console.log(`[onLeave] Client left the room: ${player.loginName}`);
+    if (player !== undefined) {
+      if (player.loginName === this.state.hostLoginName) {
+        this.broadcast(MessageType.LEFT_MESSAGE, {
+          type: MessageType.LEFT_MESSAGE,
+          message: `The Host: ${
+            this.state.playerList[player.loginName].displayName
+          } left the game.`,
+        });
+        this.state.removePlayer(player.loginName);
+        // only way to access 'first' element...
```

Comment:
We updated colysseus right? is there a cleaner way to access the first element yet ?